src/bench/peer_eviction.cpp

```diff
@@ -0,0 +1,169 @@
+// Copyright (c) 2021 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <bench/bench.h>
+#include <net.h>
+#include <netaddress.h>
+#include <random.h>
+#include <test/util/net.h>
+#include <test/util/setup_common.h>
+
+#include <algorithm>
+#include <functional>
+#include <vector>
+
+std::vector<NodeEvictionCandidate> GetRandomNodeEvictionCandidates(int num_peers, FastRandomContext& random_context)
+{
+    std::vector<NodeEvictionCandidate> candidates;
+    for (int id = 0; id < num_peers; ++id) {
+        candidates.push_back({
+            /* id */ id,
+            /* nTimeConnected */ static_cast<int64_t>(random_context.randrange(100)),
+            /* m_min_ping_time */ std::chrono::microseconds{random_context.randrange(100)},
+            /* nLastBlockTime */ static_cast<int64_t>(random_context.randrange(100)),
+            /* nLastTXTime */ static_cast<int64_t>(random_context.randrange(100)),
+            /* fRelevantServices */ random_context.randbool(),
+            /* fRelayTxes */ random_context.randbool(),
+            /* fBloomFilter */ random_context.randbool(),
+            /* nKeyedNetGroup */ random_context.randrange(100),
+            /* prefer_evict */ random_context.randbool(),
+            /* m_is_local */ random_context.randbool(),
+            /* m_network */ ALL_NETWORKS[random_context.randrange(ALL_NETWORKS.size())],
+        });
+    }
+    return candidates;
+}
+
+static std::vector<NodeEvictionCandidate> CreateEvictionCandidates(int num_peers, std::function<void(NodeEvictionCandidate&)> candidate_setup_fn, FastRandomContext& random_context)
+{
+    std::vector<NodeEvictionCandidate> candidates{GetRandomNodeEvictionCandidates(num_peers, random_context)};
+    for (NodeEvictionCandidate& candidate : candidates) {
+        candidate_setup_fn(candidate);
+    }
+    return candidates;
+}
+
+/* Benchmarks */
+
+static void EvictionProtection1Network250Candidates(benchmark::Bench& bench)
+{
+    const int num_peers{250};
+    FastRandomContext random_context{true};
+
+    const std::vector<NodeEvictionCandidate> eviction_candidates{
+        CreateEvictionCandidates(
+            num_peers, [](NodeEvictionCandidate& c) {
+                c.nTimeConnected = c.id;
+                c.m_is_local = false;
+                if (c.id >= 130 && c.id < 240) { // 110 Tor
+                    c.m_network = NET_ONION;
+                } else {
+                    c.m_network = NET_IPV6;
+                }
+            },
+            random_context)};
+
+    bench.run([&] {
+        std::vector<NodeEvictionCandidate> candidates;
+        for (int id = 0; id < 60000; ++id) {
+            candidates = eviction_candidates;
+            ProtectEvictionCandidatesByRatio(candidates);
+        }
+    });
```

Comment:
I think [nanobench](https://nanobench.ankerl.com/reference.html#_CPPv4N6ankerl9nanobench5Bench6epochsE6size_t) is designed to make such loops unnecessary. It can provide stable results even if the test function is very fast/short, no need to loop 60k times.

Also, the above benchmarks the copying of the data which could add noise to the test. Given that the input is not huge we can create a separate copy for each iteration of the test, this consumes a few tens of megabytes which is ok. Here is an example (that also uses the de-duplicated `GetRandomNodeEvictionCandidates()` from `test/util/net.h`):

```cpp
// Copyright (c) 2021 The Bitcoin Core developers
// Distributed under the MIT software license, see the accompanying
// file COPYING or http://www.opensource.org/licenses/mit-license.php.

#include <bench/bench.h>
#include <net.h>
#include <netaddress.h>
#include <random.h>
#include <test/util/net.h>
#include <test/util/setup_common.h>

#include <algorithm>
#include <functional>
#include <vector>

static void EvictionProtectionCommon(
    benchmark::Bench& bench,
    int num_candidates,
    std::function<void(NodeEvictionCandidate&)> candidate_setup_fn,
    size_t epochs,
    uint64_t epoch_iterations)
{
    using Candidates = std::vector<NodeEvictionCandidate>;

    FastRandomContext random_context{true};

    bench.epochs(epochs).epochIterations(epoch_iterations);

    Candidates candidates{GetRandomNodeEvictionCandidates(num_candidates, random_context)};
    for (auto& c : candidates) {
        candidate_setup_fn(c);
    }

    std::vector<Candidates> copies{bench.epochs() * bench.epochIterations(), candidates};
    size_t i{0};
    bench.run([&] {
        ProtectEvictionCandidatesByRatio(copies.at(i));
        ++i;
    });
}

/* Benchmarks */

static void EvictionProtection1Network250Candidates(benchmark::Bench& bench)
{
    EvictionProtectionCommon(
        bench,
        250 /* num_candidates */,
        [](NodeEvictionCandidate& c) {
            c.nTimeConnected = c.id;
            c.m_is_local = false;
            if (c.id >= 130 && c.id < 240) { // 110 Tor
                c.m_network = NET_ONION;
            } else {
                c.m_network = NET_IPV6;
            }
        },
        10 /* epochs */,
        100 /* epoch_iterations */);
}

static void EvictionProtection3Networks050Candidates(benchmark::Bench& bench)
{
    EvictionProtectionCommon(
        bench,
        50 /* num_candidates */,
        [](NodeEvictionCandidate& c) {
            c.nTimeConnected = c.id;
            c.m_is_local = (c.id == 28 || c.id == 47); //  2 localhost
            if (c.id >= 30 && c.id < 47) {             // 17 I2P
                c.m_network = NET_I2P;
            } else if (c.id >= 24 && c.id < 28) { //  4 Tor
                c.m_network = NET_ONION;
            } else {
                c.m_network = NET_IPV4;
            }
        },
        10 /* epochs */,
        100 /* epoch_iterations */);
}

static void EvictionProtection3Networks100Candidates(benchmark::Bench& bench)
{
    EvictionProtectionCommon(
        bench,
        100 /* num_candidates */,
        [](NodeEvictionCandidate& c) {
            c.nTimeConnected = c.id;
            c.m_is_local = (c.id >= 55 && c.id < 60); //  5 localhost
            if (c.id >= 70 && c.id < 80) {            // 10 I2P
                c.m_network = NET_I2P;
            } else if (c.id >= 80 && c.id < 96) { // 16 Tor
                c.m_network = NET_ONION;
            } else {
                c.m_network = NET_IPV4;
            }
        },
        10 /* epochs */,
        100 /* epoch_iterations */);
}

static void EvictionProtection3Networks250Candidates(benchmark::Bench& bench)
{
    EvictionProtectionCommon(
        bench,
        250 /* num_candidates */,
        [](NodeEvictionCandidate& c) {
            c.nTimeConnected = c.id;
            c.m_is_local = (c.id >= 140 && c.id < 160); // 20 localhost
            if (c.id >= 170 && c.id < 180) {            // 10 I2P
                c.m_network = NET_I2P;
            } else if (c.id >= 190 && c.id < 240) { // 50 Tor
                c.m_network = NET_ONION;
            } else {
                c.m_network = NET_IPV4;
            }
        },
        10 /* epochs */,
        100 /* epoch_iterations */);
}

// 1 disadvantaged network (Tor) with 250 eviction candidates.
BENCHMARK(EvictionProtection1Network250Candidates);

// 3 disadvantaged networks (I2P/localhost/Tor) with 50/100/250 eviction candidates.
BENCHMARK(EvictionProtection3Networks050Candidates);
BENCHMARK(EvictionProtection3Networks100Candidates);
BENCHMARK(EvictionProtection3Networks250Candidates);
```

Reply:
I tried this, and am playing with your proposal that looks good! but as before when I was trying to use these options, nanobench kept saying it was unstable and to increase the iterations, beyond even the numbers I settled on in my manual loops that yielded slowish but stableish results for me.
```
|            9,132.53 |          109,498.74 |    9.8% |      0.10 | :wavy_dash: `EvictionProtection3Networks050Candidates` (Unstable with ~1,000.0 iters. Increase `minEpochIterations` to e.g. 10000)
|           21,737.81 |           46,002.80 |    6.8% |      0.22 | :wavy_dash: `EvictionProtection3Networks100Candidates` (Unstable with ~1,000.0 iters. Increase `minEpochIterations` to e.g. 10000)
```
So I'm still looking at this and may punt on using the options.  

Yes, I benched the data copying alone, and it seemed insignificant relative to the call under test, not worth messing with by making (potentially many, e.g. 100k) copies instead.  That could be ok if we don't need so many copies, so I understand why you proposed reducing the number of iters.

Reply:
Here are some numbers:

* `sizeof(NodeEvictionCandidate)` is 64
* number of candidates is 250
* the current test contains 60k iterations

So that would make 64 * 250 * 60000 = 916 MiB.

In the changes above I suggested 10 epochs * 100 iterations (1000 copies): 64 * 250 * 1000 = 15 MiB.

> so I understand why you proposed reducing the number of iters

Yes, I was concerned about the memory usage, but also I stopped getting the "unstable results" warning at around 500 iters so I decided to double it to 1000 just in case, since 15 MiB is not too much. I guess a few times more than that would also be ok.

Reply:
Two random thoughts:

1. Do you get "unstable results" warnings when running other benchmarks from `src/bench`?
2. Nanobench's [`warmup()`](https://nanobench.ankerl.com/reference.html#_CPPv4N6ankerl9nanobench5Bench6warmupE8uint64_t) may help?

Reply:
I've tuned my system for benchmarking and any iterations less than in my original proposal gives errors above 5%, often 10-25%, rendering the benchmark useless for me.  I'm going to consider that my system is unusually slow and unsuited to running these and use this feedback provided that reviewers find the results stable for them.

Reply:
I've tuned my system for benchmarking, and nevertheless, any iterations less than in my original proposal still return bench error rates above 5%, often 10-25%, rendering the benchmark somewhat useless for me.

I'm going to consider that my system (2 physical cores, 4 logical ones, with an open source coreboot) is unusual and unsuited to running these benchmarks and use this feedback provided that reviewers find the results stable for them.